code/modules/projectiles/gun.dm

```diff
@@ -275,35 +279,53 @@
 		return
 
 /obj/item/weapon/gun/attackby(obj/item/I, mob/user, params)
-	if(can_flashlight)
-		if(istype(I, /obj/item/device/flashlight/seclite))
-			var/obj/item/device/flashlight/seclite/S = I
-			if(!gun_light)
-				if(!user.transferItemToLoc(I, src))
-					return
-				to_chat(user, "<span class='notice'>You click [S] into place on [src].</span>")
-				if(S.on)
-					set_light(0)
-				gun_light = S
-				update_icon()
+	if(istype(I, /obj/item/device/flashlight/seclite))
+		if(!can_flashlight)
+			return ..()
+		var/obj/item/device/flashlight/seclite/S = I
+		if(!gun_light)
+			if(!user.transferItemToLoc(I, src))
+				return
+			to_chat(user, "<span class='notice'>You click [S] into place on [src].</span>")
+			if(S.on)
+				set_light(0)
+			gun_light = S
+			update_icon()
+			update_gunlight(user)
+			verbs += /obj/item/weapon/gun/proc/toggle_gunlight
+			var/datum/action/A = new /datum/action/item_action/toggle_gunlight(src)
+			if(loc == user)
+				A.Grant(user)
+	if(istype(I, /obj/item/weapon/kitchen/knife))
+		if(!can_knife)
+			return ..()
+		var/obj/item/weapon/kitchen/knife/K = I
+		if(!bayonet)
+			if(!user.transferItemToLoc(I, src))
+				return
+			to_chat(user, "<span class='notice'>You attach [K] to the front of [src].</span>")
```

Comment:
Set bayonet somewhere.
Also, if you let any guns which can have a suppressor have a bayonet, sprites will get wonky.